python/tchannel/tornado/connection.py

```diff
@@ -618,21 +611,19 @@ def send_request(self, request):
         return response_future
 
     def adapt_result(self, f, request, response_future):
+        if not response_future.running():
+            return
+
         if f.exception():
             protocol_exception = f.exception()
             protocol_exception.tracing = request.tracing
             response_future.set_exception(protocol_exception)
-            # event: after_receive_protocol_error
-            self.tchannel.event_emitter.fire(
-                EventType.after_receive_error,
-                protocol_exception,
-            )
+
         else:
             response = f.result()
             response.tracing = request.tracing
             response_future.set_result(response)
-            # event: after_receive_response
-            self.tchannel.event_emitter.fire(
-                EventType.after_receive_response,
-                response,
-            )
+
+    def remove_outstanding_request(self, request):
+        """Remove request from pending request list"""
+        self._outstanding.pop(request.id, None)
```

Comment:
Also, depending on where else you `pop` the future, it may be easier to just add a done_callback on the Future that pops it from `_outstanding` regardless. Then you never have to manually have it evicted. When it finished, it leaves.